Added two commands to better understand the coverage of the identified functions

balzarot opened the pull request:
aau : if aai prints info about the functions identified during the
       analysis, aau prints a list of the bytes that have not been
       identified.
 afM : prints a map (inspired by the color bar on top of IDA) that
       for each byte tells you if it belongs to a function or not.

Example:

    [0x00400440]> afM 
    0x00400440  F=======....==============================......F===============
    0x00400480  =========================.......................................
    0x004004c0  ................................................................
    0x00400500  .............................................F==================
    0x00400540  ===========================================F====================
    0x00400580  ==============================================..................
    0x004005c0  ................................................................
    0x00400600  ..................................
    212 / 482 (43.98%) bytes assigned to a function
    
    [0x00400440]> aau
    0x00400448       4     main    
    0x00400499     148
    0x004005ae     116

crowell commented:
please use tabs instead of spaces.

crowell commented:
imho this belongs under `pz` because it doesn't actually do any "analysis" per-se. but I'll defer to @radare for this.

crowell commented:
needs a usage example/help in the help_msg (line 1056)

crowell commented on libr/core/cmd_anal.c:
```diff
@@ -605,6 +605,123 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+    r_list_foreach (core->anal->fcns, iter, fcn) {
+        // for each basic block in the function
+        r_list_foreach (fcn->bbs, iter2, b) { 
+            // if it is not withing range, continue
```
s/withing/within

crowell commented on libr/core/cmd_anal.c:
```diff
@@ -605,6 +605,123 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+            if ((fcn->addr < base_addr) || (fcn->addr >= base_addr+code_size))
+                continue;
+            // otherwise mark each byte in the BB in the bitmap
+            int counter = 1;
```
move this to the top of the function c89 style.

crowell commented on libr/core/cmd_anal.c:
```diff
@@ -605,6 +605,123 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+            if ((fcn->addr < base_addr) || (fcn->addr >= base_addr+code_size))
+                continue;
+            // otherwise mark each byte in the BB in the bitmap
+            int counter = 1;
```
move to initiailzers at the top of the function

radare commented:
Those commands are already implemented under the p=. Can you give them a look to avoid dupes?

> On 15 Apr 2016, at 20:02, Davide Balzarotti <notifications@github.com> wrote:
> 
> aau : if aai prints info about the functions identified during the
> analysis, aau prints a list of the bytes that have not been
> identified.
> afM : prints a map (inspired by the color bar on top of IDA) that
> for each byte tells you if it belongs to a function or not.
> 
> Example:
> 
> [0x00400440]> afM 
> 0x00400440  F=======....==============================......F===============
> 0x00400480  =========================.......................................
> 0x004004c0  ................................................................
> 0x00400500  .............................................F==================
> 0x00400540  ===========================================F====================
> 0x00400580  ==============================================..................
> 0x004005c0  ................................................................
> 0x00400600  ..................................
> 212 / 482 (43.98%) bytes assigned to a function
> 
> [0x00400440]> aau
> 0x00400448       4     main    
> 0x00400499     148
> 0x004005ae     116
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/radare/radare2/pull/4648
> 
> Commit Summary
> 
> Added two commands:
> File Changes
> 
> M libr/core/cmd_anal.c (123)
> Patch Links:
> 
> https://github.com/radare/radare2/pull/4648.patch
> https://github.com/radare/radare2/pull/4648.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
>

balzarot commented:
@crowell thanks for the comment and sorry for the style.. It's my first time time on radare
@radare I had a look at p= but i could not find anything related:

    | p=   print bytes of current block in bars
    | p=b  same as above
    | p=d  print different bytes from block
    | p=e  print entropy for each filesize/blocksize
    | p=p  print number of printable bytes for each filesize/blocksize

But again I'm a newbie so if I missed something please let me know.

Maijin commented:
What about the existing `p-`, `p-h`, `p-j`

balzarot commented:
**p-** and **p-j** are indeed related to **afM** (but not to **aau**).
**p-j** maybe it is good the way it is, but I find p- very cumbersome to use, and I think it would be
easier if the output was formatted like in **afM**.
A couple of comments:

1. I had to check the code because the help did not really help much in understanding the output
2. *p-* is a zoom-like function. To print a complete map you need to set the block to the size of the 
    section. In this case, the output is formatted as a gigantic single line, making it hard to filter (grep,
    head) and very hard to know to where each character refer to.
3. *p-* markers can collide (especially when grouped in zoom mode), making the graph hard  to read.
    For example, if in one block you have one string and one function, you would only see **z**
4.  _p-_ does not work with @ or seek. So **p- @ 0x41000** seems to ignore the offset. Also, for some 
     values of the address I get a core dump (I will try to find a small reproducible example and file a bug
    report).

Overall, imho **aau** is still useful. **p-j** could be very useful for programmatic purposes.
**afM** could be kept the way it is or somehow extended to also show flags and strings like **p-** (which at the moment I found very very hard to use, but maybe because I did not understand it right).

As a comparison, this is the output for a small program of the two commands:

    [0x00400440]> afM
    0x00400440  F=========================================......F===============
    0x00400480  =========================.......................................
    0x004004c0  .........................................F==================....
    0x00400500  .............................................F==================
    0x00400540  ===========================================F====================
    0x00400580  ==============================================..................
    0x004005c0  ................................................................
    0x00400600  ..................................
    231 / 482 (47.93%) bytes assigned to a function
   
    [0x00400440]> p-
    0x400000 [^..........................................................................................................
    ............................................................................................................................
    ........................f] 0x601048



Anyway.. my 2cents.

radare commented:
I agree with all your comments. This is a poor-man implementation and the help is not really clear either. In r2land usability is treated as a bug, so i'm open to changes if that makes sense, which is the case.

Feel free to update the PR and adopt the changes you are suggesting. The p-j command is used in the webui... /enyo only iirc


> On 16 Apr 2016, at 23:48, Davide Balzarotti <notifications@github.com> wrote:
> 
> p- and p-j are indeed related to afM (but not to aau).
> p-j maybe it is good the way it is, but I find p- very cumbersome to use, and I think it would be
> easier if the output was formatted like in afM.
> A couple of comments:
> 
> I had to check the code because the help did not really help much in understanding the output
> p- is a zoom-like function. To print a complete map you need to set the block to the size of the section. In this case, the output is formatted as a gigantic single line, making it hard to filter (grep, head) and very hard to know to where each character refer to.
> p- markers can collide (especially when grouped in zoom mode), making the graph hard to read. For example, if in one block you have one string and one function, you would only see z
> p- does not work with @ or seek. So p- @ 0x41000 seems to ignore the offset. Also, for some values of the address I get a core dump (I will try to find a small reproducible example and file a bug report).
> Overall, imho aau is still useful. p-j could be very useful for programmatic purposes.
> afM could be kept the way it is or somehow extended to also show flags and strings like p- (which at the moment I found very very hard to use, but maybe because I did not understand it right).
> 
> As a comparison, this is the output for a small program of the two commands:
> 
> [0x00400440]> afM
> 0x00400440  F=========================================......F===============
> 0x00400480  =========================.......................................
> 0x004004c0  .........................................F==================....
> 0x00400500  .............................................F==================
> 0x00400540  ===========================================F====================
> 0x00400580  ==============================================..................
> 0x004005c0  ................................................................
> 0x00400600  ..................................
> 231 / 482 (47.93%) bytes assigned to a function
> 
> [0x00400440]> p-
> 0x400000 [^..........................................................................................................
> ............................................................................................................................
> ........................f] 0x601048
> Anyway.. my 2cents.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

radare commented:
ping

balzarot commented:
Sorry, I'm traveling for a conference. I will update it tomorrow.

radare commented:
👍

> On 18 Apr 2016, at 20:51, Davide Balzarotti <notifications@github.com> wrote:
> 
> Sorry, I'm traveling for a conference. I will update it tomorrow.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

balzarot commented:
Ok, I fixed the previous errors (check commit message).
I did not touch **p-** which may still need some refactoring but that is a separate issue.

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+
+
+	// Now we print the list of memory regions that are not assigned to a function
+	chunk_size   = 0;
```
do not indent like that

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+			if ((fcn->addr < base_addr) || (fcn->addr >= base_addr+code_size))
+				continue;
+			// otherwise mark each byte in the BB in the bitmap
+			int counter = 1;
```
do not define vars in the middle of nowhere

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+			bitmap[fcn->addr-base_addr] = 'F';
+		}
+	}
+
```
too spacy

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
-            r_cons_printf ("0x%08"PFMT64x"  %6d\n", base_addr+chunk_offset, chunk_size);
-    }
-    free(bitmap);
+	RListIter *iter, *iter2;                     
```
sort vars by strlen (looks better)

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+	ut64 code_size = r_num_get (core->num, "$SS");
+	ut64 base_addr = r_num_get (core->num, "$S"); 
+	char* bitmap = calloc(1,code_size+64);
+	ut64 i;
```
put i in the same line as chunk_size

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+	RAnalBlock *b;
+	ut64 code_size = r_num_get (core->num, "$SS");
+	ut64 base_addr = r_num_get (core->num, "$S"); 
+	char* bitmap = calloc(1,code_size+64);
```
wrong indentation and spacing

radare replied:
check if code_size is < 1 after calloc and check null after bitmap

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+	// Now we print the list of memory regions that are not assigned to a function
+	chunk_size   = 0;
+	chunk_offset = 0;
+	for(i=0; i<code_size; i++){
```
missing spaces around "(" and "="

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+	}
+	if (chunk_size >= 16){
+		fcn = r_anal_get_fcn_in (core->anal, base_addr+chunk_offset, R_ANAL_FCN_TYPE_FCN | R_ANAL_FCN_TYPE_SYM);
+		if (fcn)
```
use brackets or a single r_cons_printf statement

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+		for(j=0; j<64; j++){
+			if ((i+j)==code_size) break;
+			if (bitmap[i+j]) assigned++;
+			if (show_color){
```
wrong spacing before brackets

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+	// print the bitmap
+	assigned = 0;
+	r_cons_printf ("\n address    0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef");
+	for(i=0; i<code_size; i+=64){
```
missing spaces

radare replied:
you probably want to honor scr.cols like all hexdumps do and move this code into the r_print api

balzarot replied:
Regarding the src.cols I though about that. However, since here you have a one byte per byte instead of the two per byte in hexdump, it would be very inefficient with the default value. And if the user change the cols for this output, than the hexdump will be way too large.


I dont understand the point of moving it under the r_print part. This is a custom code to print this bitmap, which is only used here. What is the advantage of moving this code to the API?

radare commented on libr/core/cmd_anal.c:
```diff
@@ -3873,6 +3866,7 @@ static int cmd_anal_all(RCore *core, const char *input) {
 		"aat", " [len]", "analyze all consecutive functions in section",
 		"aap", "", "find and analyze function preludes",
 		"aav", " [sat]", "find values referencing a specific section or map",
+		"aau", " [len]", "list mem areas (larger than len bytes) not covered by functions",
```
aa subcommands are suposed to analyze, not display analysis results

balzarot replied:
I chose **aau** because I thought that it was kind of the counterpart of **aai** (one providing info about
the analyzed bytes and one about the ones that are not). However, I see your point about **aa**
actually doing the analysis. So, should it be something like **afu**?

radare replied:
I would vote for ai subcommands. And yep that aai must be aia probably

> On 19 Apr 2016, at 21:50, Davide Balzarotti <notifications@github.com> wrote:
> 
> In libr/core/cmd_anal.c:
> 
> > @@ -3873,6 +3866,7 @@ static int cmd_anal_all(RCore *core, const char *input) {
> >  		"aat", " [len]", "analyze all consecutive functions in section",
> >  		"aap", "", "find and analyze function preludes",
> >  		"aav", " [sat]", "find values referencing a specific section or map",
> > +		"aau", " [len]", "list mem areas (larger than len bytes) not covered by functions",
> I chose aau because I thought that it was kind of the counterpart of aai (one providing info about
> the analyzed bytes and one about the ones that are not). However, I see your point about aa
> actually doing the analysis. So, should it be something like afu?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

balzarot replied:
I agree.. 
but problem is, now **ai** is used to print information about an address (page permissions, etc.)
which btw should probably not be under *a* since it is not related to analysis.

radare replied:
There are no tests for ai so we can change that behaviour withiut breaking travis. We can make current ai keep doing this and add subcommands for your stuff

> On 20 Apr 2016, at 21:01, Davide Balzarotti <notifications@github.com> wrote:
> 
> In libr/core/cmd_anal.c:
> 
> > @@ -3873,6 +3866,7 @@ static int cmd_anal_all(RCore *core, const char *input) {
> >  		"aat", " [len]", "analyze all consecutive functions in section",
> >  		"aap", "", "find and analyze function preludes",
> >  		"aav", " [sat]", "find values referencing a specific section or map",
> > +		"aau", " [len]", "list mem areas (larger than len bytes) not covered by functions",
> I agree.. 
> but problem is, now ai is used to print information about an address (page permissions, etc.)
> which btw should probably not be under a since it is not related to analysis.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+		}
+	}
+	r_cons_printf ("\n%d / %d (%.2lf%%) bytes assigned to a function\n", assigned, code_size, 100.0*((float)assigned)/code_size);
+	free(bitmap);
```
missing space before parenthesis

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
+				r_cons_printf ("%c", bitmap[i+j] ? bitmap[i+j] : '.' );
+		}
+	}
+	r_cons_printf ("\n%d / %d (%.2lf%%) bytes assigned to a function\n", assigned, code_size, 100.0*((float)assigned)/code_size);
```
possible div by zero

radare commented on libr/core/cmd_anal.c:
```diff
@@ -606,120 +607,112 @@ static int anal_fcn_add_bb(RCore *core, const char *input) {
-    }
-    r_cons_printf ("\n%d / %d (%.2lf%%) bytes assigned to a function\n", assigned, code_size, 100.0*((float)assigned)/code_size);
-    free(bitmap);
+	ut64 code_size = r_num_get (core->num, "$SS");
```
same as above. this code looks quite copypasta, so i guess we can generalize this info r_print

radare commented:
can you squash the commits into a single one?

ret2libc commented on libr/core/cmd_anal.c:
```diff
@@ -47,6 +47,7 @@ static void var_help(RCore *core, char ch) {
 		"afvg", " [idx] [addr]", "define var get reference",
 		"afvs", " [idx] [addr]", "define var set reference",
 		"afx", "[-] [from] [to]", "manipulate function xrefs",
+		"afM", "", "print byte map of functions coverage",
```
Isn't it in the wrong subcommand? To show this I have to do `afv?` which should be for variable and argument commands.

radare replied:
yeah, better put all those under `ai`

balzarot replied:
Yes, we already discusses about moving the commands (including aai) into ai.
However, I think that should be a different pr as it may affect several commands.

radare commented:
cant merge.

radare commented:
can you rebase to master? the commit seems to be broken

radare commented:
you can push -f after rebase and the PR will be updated

balzarot commented:
I rebased to upstream/master and pushed. Now the pr is closed. Is it correct?

radare commented:
no it is not, you closed the pr, i cant open it because it complains there are no commits in this branch

balzarot commented:
Very strange... I just fetched, rebased, and committed. 
But if I clone the official radare, I see that my commands are in. Are you sure you did not merge it?
That could explain why after I rebased the pr disappeared (since they two repo are in sync)

crowell commented:
/me points the blame at the pancake https://github.com/radare/radare2/commit/d1290db7902862021e861068a789baadf9320f9f <--

radare commented:
Oops yes agree i merged it yesterday. it was not inteded, please update the master with your updated changes. sorry

balzarot commented:
Ok :)
You merged after I squashed my commits together, so everything looks fine.

radare commented:
👍 thanks for confirmation. You can work on the ai command refactoring

> On 23 Apr 2016, at 08:52, Davide Balzarotti <notifications@github.com> wrote:
> 
> Ok :)
> You merged after I squashed my commits together, so everything looks fine.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>
